automatic/nexus-repository/tools/helpers.ps1

```diff
@@ -102,6 +172,359 @@ function Wait-NexusAvailability {
+      $ArchiveArgs = @{
+        Path = Get-ChildItem -Path (Join-Path $DataDir "nexus3/db/$Database/*") -Exclude "cache.stt" -File
+        DestinationPath = Join-Path $DestinationPath "$($Database.ToLower())$BaseName.zip"
+        CompressionLevel = 'NoCompression'  # Not efficient, but sufficient for this migration.
+      }
+      if ($PSCmdlet.ShouldProcess($Database, "Creating Backup of Database")) {
+        Compress-ArchiveCompat @ArchiveArgs
+        Rename-Item $ArchiveArgs.DestinationPath -NewName "$($Database.ToLower())$BaseName" -PassThru:$PassThru
+      }
+      Pop-Location
+    }
+  } finally {
+    $ProgressPreference = $PreviousProgress
+    if ($Service.Status -eq 'Running') {
+      if ($PSCmdlet.ShouldProcess($ServiceName, "Starting Service")) {
+        Start-Service $ServiceName
+      }
+    }
   }
 }
+
+function Compress-ArchiveCompat {
+  <#
+    .Synopsis
+      A drop in for creating an archive on PowerShell versions that don't have Compress-Archive
+
+    .Notes
+      Specifically written to be able to create Nexus database backups,
+      this probably shouldn't be used for other stuff.
+
+    .Example
+      $ArchiveArgs = @{
+        Path = Get-ChildItem (Join-Path $DataDir "nexus3/db/$Database/*") -File -Recurse
+        DestinationPath = Join-Path $DestinationPath "SomeName.zip"
+        CompressionLevel = 'NoCompression'
+      }
+      Compress-Archive @ArchiveArgs
+  #>
+  [CmdletBinding()]
+  param(
+    [Parameter(Mandatory, ValueFromPipeline)]
+    [System.IO.FileInfo[]]$Path,
+
+    [Parameter(Mandatory)]
+    [string]$DestinationPath,
+
+    [ValidateSet("Fastest", "NoCompression", "Optimal")]
+    [string]$CompressionLevel = "Optimal"
+  )
+  begin {
+    Add-Type -Assembly System.IO.Compression, System.IO.Compression.FileSystem
+    if (-not (Test-Path (Split-Path $DestinationPath))) {$null = mkdir (Split-Path $DestinationPath) -Force}
+    $Archive = [System.IO.Compression.ZipFile]::Open($DestinationPath, [System.IO.Compression.ZipArchiveMode]::Update)
+  }
+  process {
+    foreach ($File in $Path) {
+      $null = [System.IO.Compression.ZipFileExtensions]::CreateEntryFromFile(
+        $Archive,
+        $File.FullName,
+        "$((Resolve-Path -Path $File.FullName -Relative).TrimStart('\.'))",
+        [System.IO.Compression.CompressionLevel]::$CompressionLevel
+      )
+    }
+  }
+  end {
+    $Archive.Dispose()
+  }
+}
+
+function Test-NexusMigratorRequired {
+  <#
+    .Synopsis
+      Tests if we need to migrate this system's database
+
+    .Example
+      Test-NexusMigratorRequired -DataDir $TargetDataFolder -ProgramDir $TargetFolder
+      # Returns if a migration is required, based on the existing version and database.
+  #>
+  [OutputType([Nullable[bool]])]
+  [CmdletBinding()]
+  param(
+    # The installed version of Nexus
+    $CurrentVersion = (Get-NexusVersion -ProgramDir $ProgramDir),
+
+    # The only version of Nexus we should be upgrading from
+    $RequiredVersion = "3.70.1",
```

Comment:
Just one thing, there have been patches beyond this which will still need to be migrated from when going up to 3.71 and beyond. e.g. My Nexus instance is currently 3.70.3.1.

My suggestion would be to make this a Minimum Required Version (so test that the installed version is equal to or above without going into the 3.71.